src/words/TokTok.jl

```diff
@@ -0,0 +1,245 @@
+        "\u29d9", "\u29db", "\u29fd", "\u2e23", "\u2e25", "\u2e27", "\u2e29", "\u3009",
+        "\u300b", "\u300d", "\u300f", "\u3011", "\u3015", "\u3017", "\u3019", "\u301b",
+        "\u301e", "\u301f", "\ufd3f", "\ufe18", "\ufe36", "\ufe38", "\ufe3a", "\ufe3c",
+        "\ufe3e", "\ufe40", "\ufe42", "\ufe44", "\ufe48", "\ufe5a", "\ufe5c", "\ufe5e",
+        "\uff09", "\uff3d", "\uff5d", "\uff60")
+    
+# This is the \p{Close_Punctuation} from Perl's perluniprops             
+const CURRENCY_SYM = (
+        "\u00a2", "\u00a3", "\u00a4", "\u00a5", "\u058f", "\u060b", "\u09f2", "\u09f3", 
+        "\u0af1", "\u0bf9", "\u0e3f", "\u17db", "\u20a0", "\u20a1", "\u20a2", "\u20a3",
+        "\u20a4", "\u20a5", "\u20a6", "\u20a7", "\u20a8", "\u20a9", "\u20aa", "\u20ab",
+        "\u20ac", "\u20ad", "\u20ae", "\u20af", "\u20b0", "\u20b1", "\u20b2", "\u20b3",
+        "\u20b4", "\u20b5", "\u20b6", "\u20b7", "\u20b8", "\u20b9", "\u20ba", "\ua838",
+        "\ufdfc", "\ufe69", "\uff04", "\uffe0", "\uffe1", "\uffe5", "\uffe6", "\u09fb",)
+
+# Left/Right strip, i.e. remove heading/trailing spaces.
+const LSTRIP = (" ",) => ""           
+const RSTRIP = ("\r", "\n", "\t", "\f", "\v",) => "\n"  
+# Merge multiple spaces.
+const ONE_SPACE = ("  ",) => " "
+
+const rules_atoms = Tuple(Iterators.flatten([
+        CURRENCY_SYM,
+        FUNKY_PUNCT_1,
+        FUNKY_PUNCT_2,
+        EN_EM_DASHES,
+        PROB_SINGLE_QUOTES,
+        OPEN_PUNCT,
+        CLOSE_PUNCT    
+    ])
+
+const rules_replaces = vcat(
+        NON_BREAKING,
+        AMPERCENT,
+        TAB,
+        PIPE,
+        STUPID_QUOTES_1,
+        STUPID_QUOTES_2,
+        ONE_SPACE
+    )
+
+const rules_splitting = vcat(LSTRIP, RSTRIP)
+
+"""
+    totok_tokenize(instring::AstractString)
+This tokenizer is a simple, general tokenizer, where the input has one sentence per line; thus only final period is tokenized. 
+Tok-tok has been tested on and gives reasonably good results for English, Persian, Russian, Czech, French, German, Vietnamese,
+Tajik, and a few others.
+"""
+function toktok_tokenize(instring::AbstractString)
+    ts = TokenBuffer(instring)
+    isempty(ts.input) && return ts.tokens
+ 
+    flag = handle_final_periods(ts)
+    
+    while !isdone(ts) && ts.idx <= flag
+        url_handler1(ts) ||   # these url handlers have priority over others
+        url_handler2(ts) ||
+        url_handler3(ts) || 
+        url_handler4(ts) ||
+        atoms(ts, rules_atoms) ||
+        replaces(ts, rules_replaces) || 
+        replaces(ts, rules_splitting, boundary = true) ||
+        repeated_character_seq(ts, ",", 2) || 
+        repeated_character_seq(ts, "-", 2) || 
+        repeated_character_seq(ts, ".", 2) ||
+        number(ts) ||
+        spaces(ts) || 
+        character(ts)
+    end
+    flush!(ts)
+    return ts.tokens
+end
+
+"""
+    handle_final_periods(::TokenBuffer)
+Handles the following rules using TokenBuffer API:  
+FINAL_PERIOD_1
+FINAL_PERIOD_2
```

Comment:
Replace this with something a bit more explainy,
e.g. "handles periods that occur at the end of the input"

And then say something like "This corresponds to the `FINAL_PERIOD_1` and `FINAL_PEROD_2` rules in the original toktok perl script."